### ISSUE TITLE
Opening ipv4 listen socket on port 1883 Error: Address already in use

### ISSUE THREAD

**Issue by zjt003** [eclipse-mosquitto/mosquitto#3181]
`docker exec -it be5004c7c0cb sh`
/ # `mosquitto -v`
1733404635: mosquitto version 2.0.20 starting
1733404635: Using default config.
1733404635: Starting in local only mode. Connections will only be possible from clients running on this machine.
1733404635: Create a configuration file which defines a listener to allow remote access.
1733404635: For more details see https://mosquitto.org/documentation/authentication-methods/
1733404635: Opening ipv4 listen socket on port 1883.
1733404635: Error: Address in use
1733404635: Opening ipv6 listen socket on port 1883.
1733404635: Error: Address in use

`docker logs be5004c7c0cb`
1733404263: mosquitto version 2.0.20 starting
1733404263: Config loaded from /mosquitto/config/mosquitto.conf.
1733404263: Starting in local only mode. Connections will only be possible from clients running on this machine.
1733404263: Create a configuration file which defines a listener to allow remote access.
1733404263: For more details see https://mosquitto.org/documentation/authentication-methods/
1733404263: Opening ipv4 listen socket on port 1883.
1733404263: Opening ipv6 listen socket on port 1883.
1733404263: mosquitto version 2.0.20 running

**Comment by reubenmiller** [comment]
**General advice**: It would be more helpful to provide the full details of how you are actually running mosquitto, e.g. what docker run command are you using to start it? Given that this is an open source project, providing as much information as possible would help you get better and quicker answers to your project.

**TLDR;** This is expected since you're accidentally starting a second mosquitto process inside the container by running `mosquitto` inside a `docker exec -it` session.

**More detailed answer**
Since you haven't provide exactly how you are running mosquitto under docker, I'm assuming you've already executed the `docker run`, something like below (taken from the [docker hub](https://hub.docker.com/_/eclipse-mosquitto) page, and I just removed the volume mount for simplicity).

```sh
$ docker run -it -p 1883:1883 eclipse-mosquitto
```

After a container starts, that will start the mosquitto process. If you then open a new shell inside the container using `docker exec -it <container> sh`, and then start a second instance of mosquitto, it will fail (as you have shown) because there is already an instance of mosquitto running inside the container. The `docker logs` are showing the output of the first mosquitto instance, so that is why you don't see an error message there.

**Comment by zjt003** [comment]
> **一般建议 **：提供您实际如何运行 mosquitto 的完整详细信息会更有帮助，例如，您使用什么 docker run 命令来启动它？鉴于这是一个开源项目，提供尽可能多的信息将帮助您更好、更快地获得项目答案。
> 
> **顶级域名; ** 这是意料之中的，因为您通过运行 `mosquitto` 在 `docker exec -it` 会期。
> 
> **更详细的答案 ** 由于您没有提供在 docker 下运行 mosquitto 的确切方法，因此我假设您已经执行了 `docker run`，如下所示（取自 [docker hub ](https://hub.docker.com/_/eclipse-mosquitto) 页面，为简单起见，我刚刚删除了卷挂载）。
> 
> ```shell
> $ docker run -it -p 1883:1883 eclipse-mosquitto
> ```
> 
> 容器启动后，将启动 mosquitto 进程。如果您随后使用 `docker exec -it <container> sh`，然后启动第二个 mosquitto 实例，它将失败（如您所展示的），因为容器内已经有一个 mosquitto 实例正在运行。这 `docker logs` 显示第一个 mosquitto 实例的输出，因此您在此处看不到错误消息的原因。

`docker-compose.yml`
```yaml
services:
  mosquitto:
    image: eclipse-mosquitto
    tty: true
    container_name: mosquitto
    ports:
      - '13003:1883'
      - '13004:9001'
    volumes:
      - './config/mosquitto.conf:/mosquitto/config/mosquitto.conf'
      - './data:/mosquitto/data'
      - './log:/mosquitto/log'
```
Unable to connect through port 13003.
What methods can be used to solve this, or do we absolutely have to use port 1883?

**Comment by reubenmiller** [comment]
@zjt003 You need to provide your `mosquitto.conf` file that you're using.

But guessing, you're missing the following line to create a listener on port 1883 (inside the container):

```sh
listener 1883 0.0.0.0
```

Then afterwards, you can use normal docker port mapping to map the internal container port 1883 to the outside world port, `13003` (but you are already doing this in your docker-compose.yaml file with the line `13003:1883`

**Comment by reubenmiller** [comment]
You can see all of the mosquitto configuration file documentation here: https://mosquitto.org/man/mosquitto-conf-5.html

**Comment by zjt003** [comment]
> @zjt003 您需要提供您的 `mosquitto.conf` 文件。
> 
> 但猜想一下，您错过了以下在端口 1883（容器内）创建侦听器的代码行：
> 
> ```shell
> listener 1883 0.0.0.0
> ```
> 
> 然后，你可以使用普通的 docker 端口映射，将内部容器端口 1883 映射到外部世界端口， `13003` （但您已经在 docker-compose.yaml 文件中使用 `13003:1883`

After adding the configuration, you can connect, thank you
